notebooks/query-language-object-based-dialect/04.01.basic_numeric_comparison.ipynb

```diff
@@ -0,0 +1,574 @@
+       "\n",
+       "Based on the types of indexing (`Index or MultiIndex`), retrieving data from a DataFrame corresponding to a particular node either retrieves a single or multiple rows. This difference can cause issues when applying query node predicates.\n",
+       "Therefore, it is necessary to get rid of all index levels besides the node column through an aggregation operation on the GraphFrame. Then, a query node predicate can be applied to the GraphFrame. "
+      ],
+      "text/plain": [
+       "<IPython.core.display.Markdown object>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "# why use drop index levels\n",
+    "display(Markdown(\"../common/documentation/drop-index-information.md\"))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "f50aec9a-c6a8-420c-8d2d-0ff3d643b92b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "gf.drop_index_levels()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "06ce6daf-b98e-4006-b3d4-d56f7cd959c3",
+   "metadata": {
+    "jupyter": {
+     "source_hidden": true
+    },
+    "tags": [
+     "hide-input"
+    ]
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/markdown": [
+       "### Query type 1: Evaluate numeric metric comparison (e.g., greater than) in query node predicates\n",
+       "\n",
+       "\n",
+       "In the simplest case, a user can use this query type to find GraphFrame nodes where the time spent on the nodes is comparable to a numerical value using inequalities (<, >, <=, >=). A practical use case of such type of query is to constrain which nodes to match before conducting a numeric metric comparison. Below, we take a look at an example with a similar use case.\n",
+       "\n",
+       "The following query first filters all single nodes with the name metric `lulesh.cycle`. Then, the second query node includes all nodes below the lulesh.cycle node in the GraphFrame. The final query node matches single nodes in the GraphFrame where the numeric comparison is satisfied when the `time` metric `is greater than 100000`:"
+      ],
+      "text/plain": [
+       "<IPython.core.display.Markdown object>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "# display query type 1 documentation\n",
+    "display(Markdown(\"../common/documentation/basic-numeric-comparison-01.md\"))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "id": "83c28c99-c418-442f-bbe6-8ec3a577a817",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "query_1 = [\n",
```

Comment:
So in regards to this query, I understand we are matching all single nodes with the lulesh.cycle, and the second query includes all nodes below the lulesh.cycle node in the GraphFrame. However, for the final query, it states for time > 100,000. Wouldn't LagrangeNodal have more children showing up in it's path for the filtered tree?


Also another comment, is that the filtered tree, has different time values than the non-filtered tree. Is there a reason for this? I find it a bit odd, and if this is done purposely I would provide an explanation so the user doesn't get confused. This occurs in both filters.

Reply:
For the first question: The third quantifier, we're using a ".", which translates to "Match single nodes". Therefore, it should now make sense as to why the resulting GraphFrame looks the way it does.

For the second question: 
The reason why the values change has to do with what happens in GraphFrame.filter. filter is essentially split into 2 steps:
Apply the filter to reduce only the DataFrame
Apply GraphFrame.squash to update the Graph and columns in the DataFrame so that the GraphFrame is self-consistent
Note that step 2 is applied only when squash=True is passed to GraphFrame.filter, which is true by default.
So, for example, when using the QL, the QueryMatcher class will return a set of the Node objects captured by the query. Then, filter will remove all rows from the DataFrame that aren’t associated with the captured nodes. So, prior to step 2, the GraphFrame will have a filtered DataFrame (with no metrics changed) and the original Graph.
Then, in step 2, squash is used to prune any nodes in the Graph not found in the DataFrame. As part of this process, all inclusive metrics will be updated. To understand this update, consider the following super simple GraphFrame (i.e., one parent, A, with 3 children, B, C, and D):
A
|-  B
|-  C
|-> D
For this GraphFrame, the inclusive time of A will be the sum of A’s exclusive time and the inclusive times of A’s children (i.e., B, C, and D). Now, consider that a filter is applied that removed node B. Since the inclusive time of A includes the inclusive times of all of A’s children and since we now have one less child, it makes sense that A’s inclusive time will be less. This (along with updating the Graph object) is what squash does, and it is why the metric values are changing.
This all assumes that you’re printing an inclusive metric (e.g., time (inc)).